support/controlplane-component/defaults_test.go

```diff
@@ -0,0 +1,204 @@
+package controlplanecomponent
+
+import (
+	"sort"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	fuzz "github.com/google/gofuzz"
+)
+
+func cmVolume(name string) corev1.Volume {
+	return corev1.Volume{
+		Name: name,
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: name,
+				},
+			},
+		},
+	}
+}
+
+func secretVolume(name string) corev1.Volume {
+	return corev1.Volume{
+		Name: name,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: name,
+			},
+		},
+	}
+}
+
+func projectedCMVolume(volumeName string, cmNames ...string) corev1.Volume {
+	v := corev1.Volume{
+		Name: volumeName,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{},
+		},
+	}
+	for _, name := range cmNames {
+		v.VolumeSource.Projected.Sources = append(v.VolumeSource.Projected.Sources, corev1.VolumeProjection{
+			ConfigMap: &corev1.ConfigMapProjection{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: name,
+				},
+			},
+		})
+	}
+	return v
+}
+
+func projectedSecretVolume(volumeName string, secretNames ...string) corev1.Volume {
+	v := corev1.Volume{
+		Name: volumeName,
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{},
+		},
+	}
+	for _, name := range secretNames {
+		v.VolumeSource.Projected.Sources = append(v.VolumeSource.Projected.Sources, corev1.VolumeProjection{
+			Secret: &corev1.SecretProjection{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: name,
+				},
+			},
+		})
+	}
+	return v
+}
+
+func TestPodConfigMapNames(t *testing.T) {
+	tests := []struct {
+		name    string
+		podSpec corev1.PodSpec
+		expect  []string
+	}{
+		{
+			name: "volumes",
+			podSpec: corev1.PodSpec{
+				Volumes: []corev1.Volume{cmVolume("cm1"), secretVolume("s1"), cmVolume("cm2")},
+			},
+			expect: []string{"cm1", "cm2"},
+		},
+		{
+			name: "projected",
+			podSpec: corev1.PodSpec{
+				Volumes: []corev1.Volume{projectedCMVolume("pcms", "pcm1", "pcm2"), secretVolume("s1")},
+			},
+			expect: []string{"pcm1", "pcm2"},
+		},
+		{
+			name: "volumes and projected",
+			podSpec: corev1.PodSpec{
+				Volumes: []corev1.Volume{cmVolume("cm1"), cmVolume("cm2"), projectedCMVolume("pcms", "pcm3", "pcm4"), secretVolume("s1")},
+			},
+			expect: []string{"cm1", "cm2", "pcm3", "pcm4"},
+		},
```

Comment:
can we also add the case of empty volumes and a projectedVolume that's not neither a ConfigMap nor a Secret to verify that we skip it?